### PR TITLE
Reference-count CIDR mappings in the IPCache

### PR DESCRIPTION
**Pull request by joestringer**
```release-note
Fix bug where inserting the same CIDR in multiple rules, then removing one rule, would result in traffic not being allowed based on the rule that remains in the policy.
```

Previously, if you added two separate rules like the below, then deleted one of them, then the other rule would no longer apply because the CIDR->ID mapping was erroneously removed from the kvstore. This PR fixes that issue, and performs a bit of extra cleanup on the side. You are recommended to review this per-patch.

Policy:
```
[
  {
    "endpointSelector": {
      "matchLabels": {
        "container:id": "app1"
      }
    },
    "egress": [
      {
        "toCIDR": [
          "8.8.8.8/32"
        ]
      }
    ],
    "labels": [
      "foo"
    ]
  },
  {
    "endpointSelector": {
      "matchLabels": {
        "container:id": "app2"
      }
    },
    "egress": [
      {
        "toCIDR": [
          "8.8.8.8/32"
        ]
      }
    ],
    "labels": [
      "bar"
    ]
  }
]
```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4611)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

EDIT: Green build.

File: pkg/ipcache/kvstore.go [hunk 1]
```diff
@@ -60,7 +121,43 @@ func upsertToKVStore(ipKey string, ipIDPair identity.IPIdentityPair) error {
 		logfields.Modification: Upsert,
 	}).Debug("upserting IP->ID mapping to kvstore")
 
-	return kvstore.Update(ipKey, marshaledIPIDPair, true)
+	r.Lock()
+	defer r.Unlock()
+	refcnt, ok := r.keys[ipKey]
```
Comment (rlenglet) [inline pkg/ipcache/kvstore.go]: No need for all this logic. `r.keys[ipKey]` will return 0 if not found. So these 6 lines can be simplified into:
```
refcnt := r.keyx[ipKey] // 0 if not found
refcnt++
```

File: pkg/ipcache/kvstore.go [hunk 2]
```diff
@@ -60,7 +121,43 @@ func upsertToKVStore(ipKey string, ipIDPair identity.IPIdentityPair) error {
...
+	return err
+}
+
+// release removes a reference to the specified key. If the number of
+// references reaches 0, the key is removed from the underlying kvstore.
+func (r *kvReferenceCounter) release(key string) (err error) {
+	r.Lock()
+	defer r.Unlock()
+
+	refcnt, ok := r.keys[key]
```
Comment (rlenglet) [inline pkg/ipcache/kvstore.go]: Same here:
```
refcnt := r.keys[key] // 0 if not found
refcnt--
```

File: pkg/ipcache/kvstore.go [hunk 3]
```diff
@@ -60,7 +121,43 @@ func upsertToKVStore(ipKey string, ipIDPair identity.IPIdentityPair) error {
...
+	if ok {
+		refcnt--
+	} else {
+		refcnt = 0
```
Comment (rlenglet) [inline pkg/ipcache/kvstore.go]: That's not OK. You'll try to release they key from the store in this case, and I'm not sure whether that will be ok.
Instead, I think you should panic in this case, since this is a bug.

File: pkg/ipcache/kvstore.go [hunk 4]
```diff
@@ -161,19 +259,21 @@ func UpsertIPNetsToKVStore(prefixes []*net.IPNet, identities []*identity.Identit
 	return
 }
 
-// DeleteIPFromKVStore removes the IP->Identity mapping for the specified ip from the
-// kvstore, which will subsequently trigger an event in ipIdentityWatcher().
-func DeleteIPFromKVStore(ip string) error {
+// DeleteIPFromKVStore removes the IP->Identity mapping for the specified ip
+// from the kvstore, which will subsequently trigger an event in
+// ipIdentityWatcher().
+func DeleteIPFromKVStore(ip string) (err error) {
```
Comment (rlenglet) [inline pkg/ipcache/kvstore.go]: Revert the unnecessary change in signature: `(err error)` -> `error`.

**Review by rlenglet — APPROVED** [review]
Overall LGTM.
I think you should panic if trying to delete a non-existent key. That'd be a bug, and a panic will help us catch it.

File: pkg/ipcache/kvstore.go (hunk 3 above)
Comment (joestringer) [reply pkg/ipcache/kvstore.go]: I believe this will be okay, it will be similar to the behaviour found in #4570 - the key won't be found, the delete will fail then we'll pass the error back up to the caller and it will be logged.

Panic sounds drastic considering the consequences. I'm open to adding this to master, but it would make me nervous about whether we want to backport this PR any time soon unless we give it some good soak time.

Also, this will be dismissed by addressing your feedback immediately above.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Hit this:
```
/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:376
Unable to connect between frontend-nggmh and 10.107.202.213:80/
Expected command: kubectl exec -n qa frontend-nggmh -- curl -s -D /dev/stderr --fail --connect-timeout 5 --max-time 5 http://10.107.202.213:80/ 
To succeed, but it failed:
Exitcode: 1 
Stdout:
 	 
Stderr:
 	 error: unable to upgrade connection: container not found ("misc")
	 

/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/Policies.go:1097
```

Notably, the container isn't even ready yet:

```
STEP: Waiting for endpoints to be ready on cilium
STEP: Running tests WITHOUT Policy / Proxy loaded
STEP: running curl '10.107.202.213:80' from pod "frontend-nggmh" (should work)
=== Test Finished at 2018-06-22T06:09:32Z====
===================== TEST FAILED =====================
cmd: kubectl get pods -o wide --all-namespaces
NAMESPACE     NAME                               READY     STATUS              RESTARTS   AGE       IP              NODE
development   backend-rs2l1                      1/1       Running             0          20s       10.10.1.33      k8s2
kube-system   cilium-6w83v                       1/1       Running             0          2m        192.168.36.11   k8s1
kube-system   cilium-w1r3r                       1/1       Running             0          2m        192.168.36.12   k8s2
kube-system   etcd-k8s1                          1/1       Running             0          8m        192.168.36.11   k8s1
kube-system   kube-apiserver-k8s1                1/1       Running             0          9m        192.168.36.11   k8s1
kube-system   kube-controller-manager-k8s1       1/1       Running             0          8m        192.168.36.11   k8s1
kube-system   kube-dns-2552250787-dgvmb          3/3       Running             0          9m        10.10.1.174     k8s2
kube-system   kube-proxy-k9qc0                   1/1       Running             0          2m        192.168.36.12   k8s2
kube-system   kube-proxy-ss2c2                   1/1       Running             0          9m        192.168.36.11   k8s1
kube-system   kube-scheduler-k8s1                1/1       Running             0          8m        192.168.36.11   k8s1
kube-system   microscope                         1/1       Running             0          1m        10.10.1.2       k8s2
prometheus    prometheus-core-3506541099-2nfp8   1/1       Running             0          2m        10.10.0.150     k8s1
qa            frontend-nggmh                     0/1       ContainerCreating   0          21s       <none>          k8s1
```

I don't see any unexpected logs relating to this PR's changes.

**Comment by joestringer** [comment]
Hit #4612, will re-run as I suspect it's a flake.

**Comment by joestringer** [comment]
test-me-please

File: pkg/ipcache/kvstore.go (hunk 3 above)
Comment (tgraf) [reply pkg/ipcache/kvstore.go]: I'm absolutely against panic for anything that can reoccur. It' impossible to work around and leads to a total outage of the entire cluster if it happens based on kvstore state.

Panic is only acceptable if we can't continue and are potentially corrupting data.

Panic is not OK to force developers and users to pay attention. We should find a way to monitor errors and resolve them.

**Comment by tgraf** [comment]
Hit #4615, rebase will fix

**Comment by tgraf** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Comment by joestringer** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]
